cpu/stm32_common/periph/i2c_1.c

```diff
@@ -146,365 +147,248 @@ int i2c_release(i2c_t dev)
 {
     assert(dev < I2C_NUMOF);
 
-    uint16_t tick = TICK_TIMEOUT;
-
-    while ((i2c_config[dev].dev->ISR & I2C_ISR_BUSY) && tick--) {}
-
     periph_clk_dis(i2c_config[dev].bus, i2c_config[dev].rcc_mask);
 
     mutex_unlock(&locks[dev]);
     return 0;
 }
 
-int i2c_read_bytes(i2c_t dev, uint16_t address, void *data,
-                   size_t length, uint8_t flags)
+int i2c_write_regs(i2c_t dev, uint16_t addr, uint16_t reg,
+                   const void *data, size_t len, uint8_t flags)
 {
     assert(dev < I2C_NUMOF);
-
-    I2C_TypeDef *i2c = i2c_config[dev].dev;
-
-    if (flags & I2C_ADDR10) {
+    if (flags & (I2C_NOSTOP | I2C_NOSTART)) {
         return -EOPNOTSUPP;
     }
 
-    int ret = 0;
-
-    if (!(flags & I2C_NOSTART)) {
-        DEBUG("[i2c] read_bytes: send start condition\n");
-        /* start reception and send slave address */
-        ret = _start(i2c, address, length, I2C_FLAG_READ, flags);
-        if (ret < 0) {
-            _stop(i2c);
-            return ret;
-        }
+    I2C_TypeDef *i2c = i2c_config[dev].dev;
+    assert(i2c != NULL);
+    DEBUG("[i2c] write_regs: Starting\n");
+    /* As a higher level function we know the bus should be free */
+    if (i2c->ISR & I2C_ISR_BUSY) {
+        return -EAGAIN;
     }
-
-    DEBUG("[i2c] read_bytes: read the data\n");
-    /* read the data bytes */
-    ret = _read(i2c, data, length);
+    /* First set ADDR and register with no stop */
+    /* No RELOAD should be set so repeated start is valid */
+    int ret = _write(i2c, addr, &reg, (flags & I2C_REG16) ? 2 : 1,
+                     flags | I2C_NOSTOP, I2C_CR2_RELOAD);
     if (ret < 0) {
-        _stop(i2c);
-        DEBUG("[i2c] read_bytes: error while reading\n");
         return ret;
     }
-
-    if (!(flags & I2C_NOSTOP)) {
-        DEBUG("[i2c] read_bytes: end transmission\n");
-        ret = _stop(i2c);
-        if (ret < 0) {
-            return ret;
-        }
-    }
-
-    return ret;
+    /* Then get the data from device */
+    return _write(i2c, addr, data, len, I2C_NOSTART, 0);
 }
 
-int i2c_read_regs(i2c_t dev, uint16_t address, uint16_t reg, void *data,
-                  size_t length, uint8_t flags)
+int i2c_read_bytes(i2c_t dev, uint16_t address, void *data,
+                   size_t length, uint8_t flags)
 {
-    assert(dev < I2C_NUMOF);
+    assert(dev < I2C_NUMOF && length < 256);
```

Comment:
I was thinking of having a define for 256. To avoid hard coded values in the code and to have it defined once.

Reply:
Yup, makes sense, on it.